VolList.cpp

```diff
@@ -1,80 +1,70 @@
 #include "VolList.h"
-
 #include "OP2Memory.h"
 #include "GlobalDefines.h"
+#include <utility>
 
-VolList::VolList()
-{
-	volSearchEntryList = buffer;
-}
-
-VolList::~VolList() { }
 
 void VolList::AddVolFile(std::string volPath)
 {
-	if (IsFull()) {
-		PostErrorMessage("VolList.cpp", __LINE__, "Too many vol files loaded. Ignoring the vol file '" + volPath + "'");
-		return;
-	}
-	
-	volPaths.push_back(std::vector<char>(volPath.c_str(), volPath.c_str() + volPath.size() + 1u));
-	
-	InitializeVolSearchEntry(&volPaths.back()[0]);
-
 	OutputDebug("Add file to VolList: " + volPath + "\n");
+	volPaths.push_back(std::move(volPath));
 }
 
+// Patch reference to the original VolSearchEntry[] in Outpost2.exe to point to a replacement
+// Note: Addresses of the original array (at various offsets) are hardcoded into several instructions
 void VolList::LoadVolFiles()
 {
-	EndList();
+	std::size_t volEntryListSize = CreateVolSearchEntryList();
 
-	VolSearchEntry *vol = &volSearchEntryList[0];
-	int *vol2 = &volSearchEntryList[0].unknown1;
-	int *vol3 = &volSearchEntryList[volFileCount].unknown1;
-	VolSearchEntry *vol4 = &volSearchEntryList[volFileCount - 1];
+	// Addresses at the start of the array are used for loop initial conditions
+	auto* arrayStart1 = &volSearchEntryList[0];
+	auto* arrayStart2 = &volSearchEntryList[0].unknown1;
+	// Addresses at the end of the array are used for loop termination conditions
+	auto* arrayEnd1 = &volSearchEntryList[volEntryListSize - 1];
+	auto* arrayEnd2 = &volSearchEntryList[volEntryListSize].unknown1;

```

Comment:
I suspect this is a bug. It is reading a value past the end of the container. Perhaps it should read:

```suggestion
 	auto* arrayEnd2 = &volSearchEntryList[volEntryListSize - 1].unknown1;
```

Reply:
I believe it's supposed to be 1 past the end, much like how the loop termination condition is typically 1 past the end. I'll add more details a bit later.

Reply:
Actually, now that I'm looking at the assembly, there is a bug. Both end entries should use `volEntryListSize`. There should be no `- 1` on either of them.

Reply:
Explanation:

When you write a for loop to iterate over an array or vector, the structure is:
```cpp
for (size_t i = 0; i < someVector.size(); ++i) {
  // ... use someVector[i]
}
```

You'll notice the loop termination condition uses `i < someVector.size()`, even though `.size()` is 1 past the end. The indexes only go from `0`..`(someVector.size() - 1)`.

Now in the above example, the only use of `i` in the loop body, is to index into `someVector`, with no index adjustments, and no raw uses of `i` in a numeric expression. That is, this loop is an ideal candidate to convert to a range based for loop:
```cpp
for (auto& someElement : someVector) {
  // ... use someElement
}
```

In a range based for loop, there is no loop index. It could be implemented internally with a loop index, but there is no requirement to do so. In fact, there's a good chance it won't use a loop index. It needs the element address, not the index of it. As such, it's more direct to store the address of the element. Such a loop might be implemented similar to:
```cpp
for (auto* elementAddr = &someVector[0]; elementAddr < &someVector[someVector.size()]; elementAddr++) {
  // ... use *elementAddr
}
```

You'll notice the loop termination condition is now `elementAddr < &someVector[someVector.size()]`, even though `&someVector[someVector.size()]` is an address past the end of the vector.

Similarly, if the elements were structs, and the loop body was accessing a single struct field:
```cpp
for (size_t i = 0; i < someVector.size(); ++i) {
  // ... use someVector[i].fieldX
}
```

The optimized form would then be:
```cpp
for (auto* fieldXAddr = &someVector[0].fieldX; fieldXAddr < &someVector[someVector.size()].fieldX; fieldXAddr = reinterpret_cast<typeof(fieldXAddr)>(reinterpret_cast<char*>(fieldXAddr) + sizeof(someVector::value_type))) {
  // ... use *fieldXAddr
}
```

Again, note the end address is `&someVector[someVector.size()].fieldX`, which is past the end of the vector. In fact, with the offset of `fieldX` added to it, it could actually be quite a bit beyond the end of the vector.

Note the awful code in the 3rd position of the `for` loop above for incrementing the loop. Normally `pointer++` behaves internally as `pointerValue += sizeof(*pointer)`. That is, it increments by the size of the pointed to type. Similarly `pointer + someInt` behaves internally as `pointerValue + someInt * sizeof(*pointer)`. Here the array element is a struct, so you want to increment by the struct size. The struct size is not strongly related to the field size (what the pointer points to). It's at least as big as the field it contains, and possibly not a multiple of the field size. Hence to represent that increment accurately, and in general, you'll need some ugly casting to undo the internal treatment of pointer addition.

----

What this is all saying, is that once the code is compiled, you'll often have addresses encoded in the instruction stream, rather the indexes. For loop termination conditions, those addresses will be beyond the end of the array.